flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/LogicalCorrelateToVersionedJoinRule.scala

```diff
@@ -0,0 +1,243 @@
+    var tableVersionFunction: TableVersionFunction,
+    var versionExpression: RexNode) {
+}
+
+/**
+  * Find [[TableVersionFunction]] call and run [[VersionExpressionExtractor]] on it's operand.
+  */
+class GetTableVersionFunctionCall(
+    var rexBuilder: RexBuilder,
+    var leftSide: RelNode)
+  extends RexVisitorImpl[TableVersionFunctionCall](false) {
+
+  def visit(node: RexNode): Option[TableVersionFunctionCall] = {
+    val result = node.accept(this)
+    if (result == null) {
+      return None
+    }
+    Some(result)
+  }
+
+  override def visitCall(rexCall: RexCall): TableVersionFunctionCall = {
+    if (!rexCall.getOperator.isInstanceOf[TableSqlFunction]) {
+      return null
+    }
+    val tableFunction = rexCall.getOperator.asInstanceOf[TableSqlFunction]
+    if (!tableFunction.getTableFunction.isInstanceOf[TableVersionFunction]) {
+      return null;
+    }
+    extractVersionExpression(
+      tableFunction.getTableFunction.asInstanceOf[TableVersionFunction],
+      rexCall.getOperands())
+  }
+
+  def extractVersionExpression(
+      tableVersionFunction: TableVersionFunction,
+      operands: util.List[RexNode]): TableVersionFunctionCall = {
+    if (!operands.size().equals(1)) {
+      throw new ValidationException("Table Version Function call must have exactly one argument")
+    }
+    val versionExpressionExtractor =
+      new VersionExpressionExtractor(tableVersionFunction, rexBuilder, leftSide)
+    val versionExpression: RexNode = operands.get(0).accept(versionExpressionExtractor)
+
+    new TableVersionFunctionCall(tableVersionFunction, versionExpression)
+  }
+}
+
+/**
+  * This class extracts version expression from [[TableVersionFunction]] call and converts it
+  * into an expression that can be used in join condition. Currently support for expressions
+  * is very limited and boils down to CAST(some column reference AS TIMESTAMP).
+  *
+  * More expressions could be supported, for example:
+  * CAST(some column reference AS TIMESTAMP) + INTERVAL 1 'SECOND'
+  * But that would require to offset/transform watermarks in the operator implementing versioned
+  * join.
+  */
+class VersionExpressionExtractor(
+    var tableVersionFunction: TableVersionFunction,
+    var rexBuilder: RexBuilder,
+    var leftSide: RelNode) extends RexDefaultVisitor[RexNode] {
+
+  /**
+    * Convert correlated field access into input reference.
+    */
+  override def visitFieldAccess(fieldAccess: RexFieldAccess): RexNode = {
+    val leftIndex = leftSide.getRowType.getFieldList.indexOf(fieldAccess.getField)
+    if (leftIndex < 0) {
+      throw new IllegalStateException(
+        s"Failed to find reference to field [${fieldAccess.getField}] in node [$leftSide]")
+    }
+    rexBuilder.makeInputRef(leftSide, leftIndex)
+  }
+
+  /**
+    * We only support CAST as calls at the moment.
+    */
+  override def visitCall(call: RexCall): RexNode = {
+    if (call.getType.getSqlTypeName.name() != TypeStringUtils.TIMESTAMP.key) {
```

Comment:
Don't mix up Flink's TypeStrings and Calcite's SQL type names.

Reply:
How should I check for the type?